optimum/onnxruntime/modeling_decoder.py

```diff
@@ -648,39 +703,440 @@ class ORTModelForCausalLM(ORTModelDecoder, GenerationMixin):
+                    " To use a merged decoder, past key values must be used."
+                )
+            use_merged = False
+
+        # TODO : deprecate
+        decoder_file_name = kwargs.pop("decoder_file_name", None)
+        decoder_with_past_file_name = kwargs.pop("decoder_with_past_file_name", None)
+
+        file_name = file_name or (decoder_with_past_file_name if use_cache else decoder_file_name)
+        if file_name is None:
+            decoder_path = None
+            # We use `is not False` here to include two cases: use_merged = None (in which case we auto-detect it),
+            # and use_merged = True (explicitely specified by the user)
+            if use_merged is not False:
+                try:
+                    decoder_path = ORTModelForCausalLM.infer_onnx_filename(
+                        model_id,
+                        [DECODER_MERGED_ONNX_FILE_PATTERN],
+                        argument_name=None,
+                        subfolder=subfolder,
+                        use_auth_token=use_auth_token,
+                        revision=revision,
+                    )
+                    use_merged = True
+                    file_name = decoder_path.name
+                except FileNotFoundError as e:
+                    if use_merged is True:
+                        raise FileNotFoundError(
+                            "The parameter `use_merged=True` was passed to ORTModelForCausalLM.from_pretrained()"
+                            " but no ONNX file for a merged decoder could be found in"
+                            f" {str(Path(model_id, subfolder))}, with the error: {e}"
+                        )
+                    use_merged = False
+
+            if use_merged is False:
+                pattern = DECODER_WITH_PAST_ONNX_FILE_PATTERN if use_cache else DECODER_ONNX_FILE_PATTERN
+                # exclude decoder file for first iteration
+                decoder_path = ORTModelForCausalLM.infer_onnx_filename(
+                    model_id,
+                    [r"^((?!decoder).)*.onnx", pattern],
+                    argument_name=None,
+                    subfolder=subfolder,
+                    use_auth_token=use_auth_token,
+                    revision=revision,
+                )
+                file_name = decoder_path.name
+
+            MODEL_TO_PATCH_FOR_PAST = {
+                "bloom",
+                "mpt",
+                "llama",
+                "blenderbot-small",
+                "blenderbot",
+                "opt",
+                "pegasus",
+                "bart",
+            }
+            if file_name == ONNX_DECODER_WITH_PAST_NAME and config.model_type in MODEL_TO_PATCH_FOR_PAST:
+                raise ValueError(
+                    f"{ONNX_DECODER_WITH_PAST_NAME} not supported for the following architecture : {', '.join(MODEL_TO_PATCH_FOR_PAST)}. Please re-export your model or set use_cache=False."
+                )
```

Comment:
I am not sure to understand what is happening here.

Reply:
For [theses](https://github.com/huggingface/optimum/blob/refactorization-decoder-ort/optimum/onnxruntime/modeling_decoder.py#L977) architectures, when exporting the model with a sequence length of 1, the attention masks will be generated incorrectly for sequence length greater than 1 https://github.com/huggingface/transformers/blob/0ee45906845c8d58b9bd2df5acd90e09b00047ff/src/transformers/models/bloom/modeling_bloom.py#L654 for this reason we have introduced specific model patcher for these architectures. This results in the previously exported decoder_model_with_past to generate incorrect decoder attention masks and ultimately to generate outputs which can differ from what's expected in this specific case we raise an error asking to re-export the model

Reply:
IMO we should raise an error for all archs, or instead keep the legacy support (decoder_model, decoder_with_past_model).